### DISCUSSION TITLE
How to insert coordinates, a rotation and different axis scales (x,y,z) for a block imported ?

### DISCUSSION THREAD

**Discussion by remysompayrac** [mozman/ezdxf#848] [Issues]
Hello again, I have a question I tried my best to fix on my own but didn't succeed. 
My wish is to insert coordinates in the block references of a block I imported in my script, along with a rotation angle and (if possible), modify the xscale of the block.
Yet, I couldn't find a proper solution. The script didn't show any errors but no change was detected after testing.
Here is a piece of my code. 

```Python

    sdoc = ezdxf.readfile(f'symboles_autocad/F_{largeur}.dxf')
    tdoc = ezdxf.new()

    importer.import_paperspace_layouts()
    tblock = tdoc.blocks.new('SOURCE_ENTS')
    ents = sdoc.modelspace().query('CIRCLE LINE')
    importer.import_entities(ents, tblock)
    PORTE = tdoc.blocks.new(name='PORTE')
    importer.finalize()

    # rotation du bloc

    angle_rot = np.arctan(coeff_dir)
    angle_rot = angle_rot * 180/np.pi # attention, prendre en compte le sens direct
                     
    msp.add_blockref('PORTE', (x,y), dxfattribs={
        'xscale': 1,
        'yscale': 1,
        'rotation': angle_rot   
    })
    
    tdoc.saveas(f'data_save/FENETRE_{largeur}.dxf')

```

Regards

**Comment by mozman** [comment]
Is `msp` the modelspace of the target doc?

**Reply by remysompayrac** [reply]
yes exactly `doc = ezdxf.new('R2010')
msp = doc.modelspace()
`

**Reply by mozman** [reply]
As far I see you create a new empty BLOCK "PORTE" and add them as block reference?

You have to provide a full working example, otherwise you are just wasting my time.

**Reply by remysompayrac** [reply]
here is my full code : 

```Python
doc = ezdxf.new('R2010')
msp = doc.modelspace()
# dossier des symboles autocad
data_symbol = "symboles_siragusa" 
# calques AUTOCAD (vérifier qu'il s'agit des bons calques)

fenetre = doc.layers.add(name = "FENETRE", color = 46, linetype = "CONTINUOUS")
porte = doc.layers.add(name = "PORTE", color = 46, linetype = "CONTINUOUS")
bati_cloison = doc.layers.add(name = "BATI_CLOISON", color = 155, linetype = "CONTINUOUS")
bati_int = doc.layers.add(name = "BATI_INT", color = 7, linetype = "CONTINUOUS")

# PORTE

#ATTENTION IL N'Y A QUE LES PORTES D'INTERIEUR

def porte(x,y,largeur,coeff_dir,coté):
      
    if largeur > 58 and largeur <= 68 :
        largeur = 63
    elif largeur > 68 and largeur <= 78 :
        largeur = 73
    elif largeur > 78 and largeur <= 88 :
        largeur = 83
    elif largeur > 88 and largeur <= 98 :
        largeur = 93
        
    # déterminer de quel côté se trouve l'ouvrant

    #côté peut être droite ou gauche --> côté de l'arc
    # côté = droite ou gauche
    sdoc = ezdxf.readfile(f'{data_symbol}/P{largeur}_{coté}.dxf')
    tdoc = ezdxf.new()

    ### Don't redefine the Importer class!
    importer = Importer(sdoc, tdoc)
    importer.import_modelspace()

    importer.import_paperspace_layouts()
    tblock = tdoc.blocks.new('SOURCE_ENTS')
    ents = sdoc.modelspace().query('CIRCLE LINE')
    importer.import_entities(ents, tblock)
    PORTE = tdoc.blocks.new(name='PORTE')
    importer.finalize()

    # rotation du bloc

    angle_rot = np.arctan(coeff_dir)  # numpy? why?
    angle_rot = angle_rot * 180/np.pi # attention, prendre en compte le sens direct
                     
    msp.add_blockref('PORTE', (x,y), dxfattribs={
        'xscale': largeur/1.15,
        'yscale': 1,
        'rotation': angle_rot   # <<< expects a float not a numpy array
    })

    tdoc.dxfattribs=({"layer": "PORTE"})  # <<< this does nothing
    tdoc.dxfattribs=({"rotation": "angle_rot"})  # <<< this does nothing
    tdoc.saveas(f'data_save/PORTE_{largeur}_{coté}.dxf')

```

**Reply by mozman** [reply]
That does the same as the previous code: you define an empty BLOCK "PORTE" and insert it as block reference, what do you expect to see?

**Reply by remysompayrac** [reply]
I expect to be able to import the BLOCK PORTE with a function that allow me to import it with specific coordinates (x,y) and change the xscale of the block and make it rotate. But it doesn't work

**Reply by mozman** [reply]
The simple question was: What do expect to see when you insert an EMPTY (=no content) block?

**Reply by remysompayrac** [reply]
It is not an EMPTY block because when I save the document, I can read the dxf with the block inside. 
It reads the dxf of a block I made on Autocad.
The objective of this code is to import this block and create a function to insert it into the dxf with specific features like coordinates (x,y), the scales, and a rotation angle.

The only thing that doesn't show is the change in the xscale, the rotation of the block and the coordinates of the block which don't apply.That is my only question and it has always been from the beginning. How to add these attributes to the block I just imported ?

Maybe I misunderstood your statement but I have to say it's quite unclear. 
When you say EMPTY block, do you talk about "msp" or "PORTE"?

**Reply by mozman** [reply]
You create a new block "PORTE" at (1), its empty. You add a block reference of "PORTE" at (2) it's still empty!
![image](https://user-images.githubusercontent.com/4230966/225628421-338e193d-3a81-4ccf-b91a-0cb50c1ae205.png)

**Reply by mozman** [reply]
Why are you importing the DXF drawing in the first place, just use the source document as a template, add your stuff and save a new DXF document.

1. you import the modelspace
2.  you import all paperspace layouts
3. you import some entities from the modelspace a second time
4.  you do NOT import any block definitions

**Reply by remysompayrac** [reply]
Thank you for all your explanations, I understand and I'm going to code it :)
